---
title: "Class notes"
author: "Alex Faidiga"
date: "January 20, 2020"
output: html_document
---
##Week 1: setting up git and github

* You can edit a file in R, in your computer’s terminal, or in the browser on GitHub and integrate the changes across all platforms.
* .Rmd files are RMarkdown files.

Why build phylogenies? Video

phylogenies can be used for:
*	estimating extinction/speciation/diversification
*formulating hypotheses (example of gene being predicted and induced)
* testing hypotheses (example of asking which traits are important for diversification in certain groups)
* determining whether or not there are correlations between traits (leaf size and longevity in conifers and angiosperms—correlation is not there when accounting for phylogeny)

PHLAWD
* finds genes in genbank and builds trees for you 

GBIF	
* species distribution data

##Week 3
###Tree terminology
Trees can tell us about nested relationships between organisms, but we lose a lot of biology in looking simply at a tree representation (introgression, hybridization, etc.).

**Branch** = edge

**Internal node** = where different edges meet

  * If all nodes have only two descendant branches, the tree is **bifurcating**, **fully resolved**, or **dichotomous**. Nodes with more branches are **polytomies.**
  * **soft polytomies** are when we don't know why there are >2 descendants; **hard polytomies** are when the lineage actually simultaneously split into multiple taxa.
  
**Rooted trees** have nodes that represent the  most recent common ancestor (MRCA). Rooted trees show the direction of time.
  * One way to root is to include an outgroup.
  * Why does rooting matter? Depending on where we place the root, it could have different implications for the direction of change of different traits.
  
**Clade** = ancestor and all its descendants. (**Monophyletic** group)

**Parahyletic group** = ancestor and some but not all of its descendants.
  * dinosaurs not including birds
  
**Polyphyletic group** = taxa not sharing recent common ancestor, but obviously if you go back far enough you can find one

**homology** = similarity due to shared ancestry

**homoplasy** = similarity not due to common ancestry

**plesiomorphy** = ancestral character state, in reference to another derived character state

**apomorphy** = derived character state, in reference to another ancestral character state

Why is the term "basal" problematic?
  * Calling a taxon basal (toward the left of the tree) implies that it is more primitive or less evolutionarily advanced than taxa toward the right of the tree, but phylogenies represent nested relationships, not how advanced a taxon is. If we wanted to, we could rearrange the tree with different taxa and make it so that the more "advanced" taxon is on the left. 
  
###Tree space

Why is searching for trees a difficult problem?

  * The more taxa on a tree, the more possible trees... exponential growth! after 50 taxa, there are more possible trees than there are atoms in the universe!
  
  * It's hard to find the right tree when tree space is so large
  
  * When tree space is so large, the odds of finding the exact right tree are very small. Any tree will have some level of error.
  
###Likelihood vs. Bayes

#####Coin flip

Likelihood approach:

  * first choose a model (binomial distribution)
  
  * assume that the coin has an innate probability of landing heads
  
  * we now have a model and our data 2 heads out of 3 tosses
  
  * the likelihood approach: **likelihood of the hypothesis given the data is proportional to the probability of the data given the hypothesis**
  
  * hypothesis is a given value for q, and data are 2 heads out of 3 tosses
  
  * we then plug into binomial formula, and we have the likelihood
  
  *the best estimate of q given likelihood approach is the value of q that maximizes the probability of the observed data.
  
  *so, the maximum likelihood estimate of q is 2/3, or the number of heads/tails.

Limitations of likelihood approach:

  * if we had more data, we would probably get a better estimate.
  
  * we have some innate assumption about coins (that they are fair), but we can't incorporate that into the analysis
  
  * a more natural result would be to get the prob. of the hypothesis given the data.

####Baye's Rule:

Used to obtain the posterior probability from the likelihood and prior probability, or: probability of the hypothesis given the data. 

>P(hyp|data) = P(data|hyp)P(hyp)/sum(P(data|hyp)P(hyp))

Hopefully, the more data you have, the less the prior matters.
    
**posterior probability** = Bayes' rule

###Drummond and Rambaut (2007). BEAST: Bayesian evolutionary analysis by sampling trees

Bayesian statistical framework allows for prior knowledge to be input in comination with information provided by the data.

**Markov Chain Monte Carlo**: "state of the art" method for phylogenetic reconstruction

  * stochastic algorithm that produces sample-based estimates of a target distribution of choice
  
  * target dist.: posterior distribution of a set of evolutionary parameters given a set of molecular sequences

coalescent-based population genetics of microsatellites? based on the assumption of no recombination, selection, gene flow

MCMC applied to central problem of co-estimation of phylogeny and sequence alignment?

BEAST focuses on calibrated (rooted, time scale incorporated) trees.

  * **molecular clock model**: uniform rate of molecular evolution over the entire tree
  * BEAST allows for **relaxed molecular clock models**, or models where the rate of molecular evolution is not assumed to be constant across all lineages
  
In example section, they mention that the first 500 samples were discarded as burnin?

###Lewis (2001). Phylogenetic systematics turns over a new leaf.

**Bootstrapping**: a statistical technique in which new data sets of the same size as the original are created by sampling randomly (and with replacement) from the original characters. It is assumed that, if the original n is sufficiently large, sampling from the sample population approximates the true underlying distribution. This allows us to calculate confidence intervals on phylogenies. Bootstrapping assumes that the original characters were sampled independently.

**Likelihood**: a quantity that is proportional to the probability of the data (or probability density, if the data are continuous-valued), given specific values for all parameters in the model. The likelihood function provides a means to estimate the parameters of the model. Parameter values that are associated with the global maximum ofthe likelihood function are termed maximum likelihood estimates (MLEs).

**Optimality criterion**: a rule used to decide which of two trees is best. Four optimality criteria are currently widely used: 

  * **Maximum parsimony** – the tree requiring the fewest character state changes is the better of the two trees.
  * **Maximum likelihood** – the tree maximizing the likelihood under the assumed evolutionary model is better.
  * **Minimum evolution**– the tree having the smallest sum of branch lengths (estimated using ordinary least squares) is better.
  * **Least squares** – the tree showing the best fit between estimated pairwise distances and the corresponding pairwise distances obtained by summing paths through the tree is better.


In phylogenetic applications, each step in a **Markov chain Monte Carlo (MCMC)** involves a random modification of the tree topology, a branch length or a parameter in the substitution model (e.g. substitution rate ratio). If the 
posterior that is computed for a proposed step is larger than that of the current tree topology and parameter values, the proposed step is taken.

  * this allows one to explore parameter space with complex topologies
  
  * it is mentioned several times that the topology is a parameter as well as branch lengths for each branch (2N-3 where N is the number of tips). How can topology be summarized with one parameter?

**General time-reversible model** of molecular evolution: allows unequal nucleotide frequencies as well as the six possible nucleotide transitions to occur at different rates.

**discrete gamma method** allows relative rates of nucleotide substitution to differ among sites. Relative rates are distributed according to gamma distribution where mean = 1 and variance = 1/a

**codon models** allow for the non-independence of nucleotide substitution rates within codons, and **secondary structure models** allow for non-independence among sites known to be correlated due to secondary strucure (ex. stem-loop structure in RNA that are associated with protein folding).

  * example: in genes that encode proteins, the three nucleotide sites that form a single codon cannot evolve independently of one another if there is selection for a particular amino acid at the corresponding site in the polypeptide.

Two traits are **evolutionarily correlated** if a change in state of one of the characters predisposes the other character to change state soon after. i.e., the two traits are correlated to a greater extent than expected from simply the phylogeny.

  * example: evolution of feathers is correlated with the evolution of wings. If feathers did not appear first, wings may not have evolved 
  
  * can do a likelihood ratio test to calculate this probability 
  
## In-class notes

introgression vs. hybridization
introgression = excange of genes, while hybridization = production of a new species from it 

flat prior = all parameter estimates are equally likely. For coin, q=prob(heads); could be any value. When we collect data, we get a point estimate on the value of the parameter

## Felsenstein 1978: Cases in which parsimony or compatibility methods will be positively misleading

The more data you have, the more certain you are in your wrong tree

Parsimony score = number of changes on the tree to reach a given character state for each tip

## 1/24 RAxML tutorial

I can't execute the raxml command without typing ./ first. I tried copying the raxml files to my directory like the tutorial said but it's not working. 

The raxml result files that are output cannot be read in Dendroscpe as they are-- I have to convert them to .txt files first. Is there a way to directly write them to .txt files?

**what does this mean:**
"As starting trees RAxML uses randomized stepwise addition parsimony trees, i.e., it will not generate the same starting tree every time. You can either force it by providing a fixed random number seed or by passing a starting tree. I fyou want to conduct multiple searches for the best tree you can type _______ and raxml will carry out 20 ML searches on 20 randomized stepwise addition parsimony trees."

**Under the bootstrapping section:**
"This call will produce to output files that can be visualized with Dendroscope:  RAxML_bipartitions.T15 (support values assigned to nodes) and RAxML_bipartitionsBranchLabels.T15 (support values assigned to branches of the tree). Note that, for unrooted trees the correct representation is actually the one with support values assigned to branches and not nodes of the tree!"

I'm still a little confused on why this is the case or what the difference is between the bootstrap values being on the nodes vs. branches

What is a "core" on my computer :(

**partitioning**: Partitioning is a commonly used method in phylogenetics that aims to accommodate variation in substitution patterns among sites.

When you are passing trees to RAxML it usually expects them to be bifurcating. The rationale for this is that a multifurcating tree actually represents a set of bifurcating trees and it is unclear how to properly resolve the multifurcations in general. Also, for computing the likelihood of a tree we need a bifurcating tree, therefore resolving multi-furcations, either at random or in some more clever way is a necessary prerequisite for computing likelihood scores.
  * is that actually the case, that multifurcations represent a set of bifurcating trees? or is that just what this author thinks?Also, why do we need strictly bifurcating trees to do max likelihood 
  
###1/28
####coalescence 

small group of salamanders vs group of all salamanders; both can have issues with coalescence of gene trees not matching species tree because coalesence can occur at any point in time. population size is more likely to affect it though

time to coalescense is faster at a lower effective population size 

the more generations, the greater the probability of coalescing 

anomaly zone=the majority of the gene trees have a history different from the species tree

**question**: rcoal() and rtree() both build random tree, but the help package says that rcoal() builds coalescent trees. Why are these random trees considered coalescent? because all of the tips are at one time point and merge back to a common ancestor?

bootstrapping measures the robustness of the data to resampling, but not necessarily the accuracy of the model. 

  * this is why large sample sizes are not great for bootstrapping-- you will probably get the same answer 
  
####Heled and Drummond 
simulated data with their prior (think of doing a coin flip simuation where p=0.3, then using that as a prior for a bayesian analysis... you will get the same answer)


####Hahn 2015

**adaptive introgression**: the traits that tend to get in and spread are usually adaptive 

  * this is because when a new gene enters the population, its frequency is low and tehrefore not likely to get passed on; so if it does, it's probably adaptive 
  

###2/4/20

**mutation rate**: rate at which DNA is changing
**substitution rate**: rate at which DNA is changing across an entire population 

if an allele is adaptive, the chance that it will go to fixation is higher. if maladaptove, lower. 

poisson distributions are used for discrete count data of rates.

  * rate*time. 2 bees per hour. 5 car crashes per mile. 
  
####Sanderson

attempting to solve the complex issue of different rates on different branches

this is an issue because if each branch has its own rate, there will be more parameters in the model than there are data. In their method, they allowed some branch lengths to be the same based on how extreme the difference is?

###2/6/20

parametric methods assume an underlying distribution that can be characterized with parameters (e.g. maximum likelihood)

BEAST allows parameters to be "linked", which essentially just means that you can set parameters for different genes/clades etc to be the same and some to be different.
  
RAxML doesnt have the capability to date trees, but you can use the output from it as a prior for BEAST and not allow it to change the tree topology, but allow it to change branch lengths.

**Instertion-deletion model:**
"multiple sequence alignment" under an insertion-deletion model basically means that you have mutiple sequences, and when there are "gaps" of amino acids that don't match, it could be an insterion or deletion. So you would use max likelihood to find the most probable alignment based on that model. BEAST allows you to estimate the alignment and the phylogeny simultaneously with MCMC, meaning that it will change the alignment, calculate likelihood, change the tree, calculate the likelihood, etc. until it finds the max likelihood estimate of both.



Using BEAST and using uncorrelated rate models can still cause you to incorrectly date trees 


###2/11/20

markov chain process: memoryless, doesnt matter where you came from, only where you are

For ancestral state reconstruction, we can build a rate matrix with the probability of going from one character state to another. Then, we can set different transtitions to zero (build different models) and test which model fits the data (tip character states) the best, to predict probabilities of transitioning states. 

To constrain by time, multiply the matrix by t (ex. rate of changing from A to B in 100 million years). When you exponentiate this matrix, you now have a  probability transition matrix (probability of changing from A to B in 100 million years)

From this basic rate matrix mode, we can:

  * test hypotheses about rates for a single character 
  * test hypotheses about correlations between characters (do they evolve at the same rate?)
  * infer trees
  * reconstruct ancestral character states
  
####Pagel (1999)

how do you decide the weight of two different character states at the root of the phylogeny to do ancestral state reconstruction? One way is to do the frequency of each phenotype at the tips.

###2/13
####In-class exercise: likelihood, optimization, MCMC

Likelihood optimization function: What parameters for the model optimize the likelihood of the data? We want to find the "best" value

Bayesian optimization function: Sample parameter space, and if the likelihood is better after one step you can move to that place BUT also sometimes you move down, based on the value of a random number. 

likelihood times prior or plus prior if data is on log scale 


###2/18

central limits theorem: the distribution of the sums of numbers drawn from any distribution will form a normal distribution.

  * brownian motion: random processes adding up along the branches result in a normal distribution of continuous variables at the tips 
  * all different processes added together will evenutally sum to a normal distribution
  -->even for any amount of taxa, or any distance between taxa?
  
###2/20

you can tell whether or not something is evolving under borwnian motion, but you can't tell what mechanism it is (drift vs. directional selection)

you can compare models for trees where some groups evolve faster than others, and does this correlate with a certain trait (like selfing)?

You can also compare models for trees where evolution occurs more quickly after a certain time point on the tree (ex. the cretaceous in agiosperms)
  
OU process: brownian motion + being drawn toward an optimum with some wiggle allowed 

theta = optimal value (parent)
sigma squared = wiggle (brownian motion)
alpha = strength of pull

different from directional selection because directional selection theorertically keeps selecting for one direction continually, whereas the OU process for stabilizing selection selects for some optimum

time vs. variance at the tip will be linear positive for a BM process
time vs. variance at the tip should be positive but eventually flatten out for OU process because there is an optimum, therefore constraint on variance 

###2/25

```{r}
pdf(file="~Desktop/Big.pdf", width=20, height=20) #this writes to a PDF
dev.off() #if i make another figure before this line it will write two PDFs to one file

drop.tip(phylogeny, "tip name") #drop a single tip from a phylogeny 
```

The brownian motion assumption means that more time=more change

With independent contrasts, brownian motion is assumed. If branch lengths are incorrect, it could affect the analysis


###2/27

For binary traits

In pagel's model, trait evolution is markovian (memoryless); state at time t does not depend on the state of the ancestor 

  * not biologically realistic because for example, it's probably easier for a eyeless fish to gain eyes than for a kelp to gain eyes 
  
You can have different rates for 1-->0 or 0-->1

###3/3
phylogeography is typically within species, while bioegeography is between

vicariance = allopatric speciation, basically; speciation due to a geogrphic barrier

dispersal = a barrier arising but groups dispersing across it, then speciation happening








